rust/flexbuffers/src/builder/map.rs

```diff
@@ -0,0 +1,114 @@
+/// commited to the buffer. If this map is the root of the flexbuffer, then the
+/// root is written and the flexbuffer is complete.
+/// ## Panics:
+/// -  Duplicate keys will result in a panic in both debug and release mode.
+/// -  Keys with internal nulls results in a panic in debug mode and result in silent truncaction
+///    in release mode.
+pub struct MapBuilder<'a> {
+    pub(super) builder: &'a mut Builder,
+    // If the root is this map then start == None. Otherwise start is the
+    // number of values in the 'values stack' before adding this map.
+    pub(super) start: Option<usize>,
+}
+impl<'a> MapBuilder<'a> {
+    /// Push `p` onto this map with key `key`.
+    /// This will panic (in debug mode) if `key` contains internal nulls.
+    #[inline]
+    pub fn push<P: Pushable>(&mut self, key: &str, p: P) {
+        self.builder.push_key(key);
+        self.builder.push(p);
+    }
+    /// Starts a nested vector that will be pushed onto this map
+    /// with key `key` when it is dropped.
+    ///
+    /// This will panic (in debug mode) if `key` contains internal nulls.
+    #[inline]
+    pub fn start_vector(&mut self, key: &str) -> VectorBuilder {
+        // Push the key that refers to this nested vector.
+        self.builder.push_key(key);
+        // Nested vector.
+        let start = Some(self.builder.values.len());
+        VectorBuilder {
+            builder: &mut self.builder,
+            start,
+        }
+    }
+    /// Starts a nested map which that will be pushed onto this map
+    /// with key `key` when it is dropped.
+    ///
+    /// This will panic (in debug mode) if `key` contains internal nulls.
+    #[inline]
+    pub fn start_map(&mut self, key: &str) -> MapBuilder {
+        // Push the key that refers to this nested vector.
+        self.builder.push_key(key);
+        // Nested map.
+        let start = Some(self.builder.values.len());
+        MapBuilder {
+            builder: &mut self.builder,
+            start,
+        }
+    }
+    /// `end_map` sorts the map by key and writes it to the buffer. This happens anyway
+    /// when the map builder is dropped.
+    #[inline]
+    pub fn end_map(self) {}
+}
+impl<'a> Drop for MapBuilder<'a> {
+    #[inline]
+    fn drop(&mut self) {
+        self.builder.end_map_or_vector(true, self.start);
+    }
+}
+
+// Read known keys / strings as iterators over bytes -- skipping utf8 validation and strlen.
+pub(super) fn get_key(buffer: &[u8], address: usize) -> impl Iterator<Item = &u8> {
+    buffer[address..].iter().take_while(|&&b| b != b'\0')
+}
+
+// `values` is assumed to be of the format [key1, value1, ..., keyN, valueN].
+// The keys refer to cstrings in `buffer`. When this function returns,
+// `values` is sorted in place by key.
+pub(super) fn sort_map_by_keys(values: &mut [Value], buffer: &[u8]) {
+    debug_assert_eq!(values.len() % 2, 0);
+    debug_assert!(values.iter().step_by(2).all(Value::is_key));
+    let pairs: &mut [[Value; 2]] = unsafe {
```

Comment:
I don't think so. I agree that 100% safe is a good goal but the reason I chose unsafe was to get all of the following:

1. "chunk" the slice into pairs
2. sort in place (to avoid allocations and use one `values` stack for all vectors and maps)

Rust can get you chunks but only in an iterator. We can collect that iterator and sort, but then it's no longer in place and we'd lose performance.

I haven't found crate that can do this for us and I think writing a custom, `sort_pairs_unstable` function is not worth removing the unsafe over.

Reply:
Okay, 1. please minimize the size of the unsafe block (`as_mut_ptr` is not unsafe), and 2. please document your complete rationale in a comment above the unsafe block.

Reply:
Done